### ISSUE TITLE
Reaching the first frame

### ISSUE THREAD

**Issue by Admer456** [Admer456/btx-engine#1]
To get there, we need a:
- [x] Console system, which requires: (October 2021)
- [x] console commands,
- [x] console variables,
- [x] its own printing;
- [x] The clientside, which requires:
- [x] an input system, (late October 2021)
- [x] a model manager, (September 2022)
- [x] most basic renderer (October 2022)

At that point, we can enter maps, render them and move around. Interaction will be outlined in some other issue.

**Comment by admer456-stu** [comment]
Should probably start working on the rendering stuff this week.

**Comment by admer456-stu** [comment]
I mean, the rendering stuff's gonna involve more or less everything. We first gotta have something to render.
Having a model manager means we're gonna have a generic model format, and I'll write a TrenchBroom map preprocessor to produce BTX models. (In the future, there should also be a simple GUI tool for importing models into the engine)

Only then we can think about rendering. It'll have an OpenGL 3.3 backend, but ultimately I wanna do Vulkan as well, for future-proofing and stuff.

**Comment by admer456-stu** [comment]
Also that "concept of a client and game logic" is proooobably solved by now lol.
![image](https://user-images.githubusercontent.com/103189008/162692396-a783f103-c1e4-49b3-93f1-ac5a73630dcb.png)

**Comment by Admer456** [comment]
> Also that "concept of a client and game logic" is proooobably solved by now lol. 

Edited to reflect this.

**Comment by admer456-stu** [comment]
Did the first draft for the model format specification. This will be the basis for next month's map preprocessor, model parser and model manager.

Model parsing will also require some additions to `Lexer`.

**Comment by Admer456** [comment]
Now that the engine has a plugin system, any sort of user interaction and game logic is easily delegated to a game/application module.

Work on the model manager & rendering should finally begin in a couple of weeks. Just gotta get some windowing stuff done for DirectX.

**Comment by Admer456** [comment]
Done. NVRHI is more or less ready to be the render backend here.

**Comment by Admer456** [comment]
> This will be the basis for next month's map preprocessor

I like how I keep saying "next month" and actually getting to it half a year later.

> Only then we can think about rendering. It'll have an OpenGL 3.3 backend, but ultimately I wanna do Vulkan as well, for future-proofing and stuff.

Oh and in case this isn't immediately clear, there won't be an OpenGL render backend. Now that I'm using NVRHI, I got Vulkan and DX12 practically without much effort.

But yeah, plans for the engine's 1st anniversary involve finally doing the render backend stuff and a TrenchBroom map parser for the game SDK. Gonna put up more design documents too.

**Comment by Admer456** [comment]
Updated the months to reflect when the model & material manager should be, and when the renderer should be. Gonna write the render frontend idea today.

**Comment by Admer456** [comment]
We are THIS 🤏 close to reaching the "first frame" and my iconic "hello pentagon" shape. I pushed some work regarding render model data, and the example usage. Not ready yet! I gotta implement ModelManager tomorrow and then we'll finally be able to generate vertex buffers in the renderer.

This is pretty exciting. Once this is done though, I'll be taking a break from the engine until mid November so I can focus on some other projects & college.

**Comment by admer456-stu** [comment]
Model manager can be ticked now since it's basically done.

Today I just gotta create a graphics pipeline, write a couple shaders and maybe add a flying camera in the test game module. Then we'll be able to fly around and the "first frame" will finally have been reached.

Yay.

**Comment by Admer456** [comment]
Should probably create a separate issue for the rendering stuff, or expand the todo list here. The "most basic renderer" is still quite a bit of work.

**Comment by Admer456** [comment]
This issue is now officially completed. The engine now renders untextured triangle meshes. By the end of the month we can expect some basic flying camera and TrenchBroom map loader, which will be outlined in another issue.
